### PR TITLE
Feature: stream test output progressively

### PR DESCRIPTION
**Pull request by mdkitzman**
# Progressive Test Results

This test reporter seems to blart out the test results only when done, so I wanted to update it to print the results as they are available

[Screencast from 2026-04-23 08-10-14.webm](https://github.com/user-attachments/assets/22153279-3369-48a9-9f90-0bdc01d5fa51)

## Summary
- make `@reporters/mocha` emit Mocha runner events progressively as `node:test` tests finish
- preserve existing built-in reporter snapshot output while still streaming per-test results
- restore root suite lifecycle events for empty or fully filtered runs
- add targeted test coverage for progressive output, custom reporter callbacks, concurrent completion order, and empty-source behavior

## What Changed
- replaced the end-of-run tree replay in `packages/mocha/index.js` with an active-node event bridge
- emit suite/test lifecycle events as `test:start`, `test:pass`, and `test:fail` arrive
- lazily start the synthetic root suite so default Mocha reporter formatting stays unchanged
- ensure the root suite lifecycle is still emitted for empty runs
- added test fixtures for:
  - progressive custom reporter output
  - empty-source suite lifecycle

## Testing
- `node --test tests/index.test.js` (from `packages/mocha`)
- `./node_modules/.bin/eslint packages/mocha/index.js packages/mocha/tests/index.test.js packages/mocha/tests/progressiveReporter/reporter.js packages/mocha/tests/emptySuiteReporter/reporter.js`

## Notes
- scope is limited to `@reporters/mocha`
- no repo-wide reporter behavior was changed

### CODE REVIEW

File: packages/mocha/tests/index.test.js [hunk 1]
```diff
@@ -41,3 +79,109 @@ test('empty', async () => {
 test('single test', async () => {
   await assert.doesNotReject(reporter([{ type: 'test:pass', data: { name: 'test', nesting: 0, details: { duration_ms: 100 } } }]));
 });
+
+test('emits root suite lifecycle for empty sources', async () => {
```
Comment (MoLow) [inline packages/mocha/tests/index.test.js:104]: please use the existing tests pattern using snapshots

Comment (mdkitzman) [reply packages/mocha/tests/index.test.js:104]: After updating these tests to use snapshots, the `gh` and `github` tests started to fail as well, and I had to do a snapshot update for those as well to get them passing

**Comment by mdkitzman** [comment]
@MoLow Anything else you need from me on this?

**Comment by mdkitzman** [comment]
@MoLow bump

File: packages/mocha/tests/index.test.js [hunk 2]
```diff
@@ -41,3 +100,46 @@ test('empty', async () => {
...
+    stdout: output.stdout(),
+    stderr: output.stderr(),
+  }, partialStdout);
+});
+
+test('streams custom reporter callbacks before suite completion', async () => {
+  const child = spawn(process.execPath, ['--test-reporter', '../../index.js', '../../../../tests/slow_tests.js'], { env: {},  cwd: resolve('./tests/progressiveReporter') });
+  const output = collectOutput(child);
+
+  const partialStdout = await waitForStdoutMatch(child, /test end:/);
```
Comment (MoLow) [inline packages/mocha/tests/index.test.js:128]: snapshoting this before it finishes can be very flaky, any reason not to follow the other tests pattern?

**Comment by MoLow** [comment]
@mdkitzman sorry for the delay, missed your messages. implementation looks ok but tests are broken

File: packages/mocha/tests/index.test.js (hunk 2 above)
Comment (mdkitzman) [reply packages/mocha/tests/index.test.js:128]: The intention here is to prove that test results are being streamed out prior to the completion of the test suite.  Using the end state snapshot does not indicate that the test results were progressively printed during the test run.

I didn't see any other tests in the repo that were asserting a similar behavior, so this seemed like a decent approach while still utilizing the snapshot functionality.

Any other thoughts on an approach that would be not as flaky?

**Comment by mdkitzman** [comment]
> @mdkitzman sorry for the delay, missed your messages. implementation looks ok but tests are broken

Thanks for looking over this again.  Looks like the run failed on a lint failure in the test where I had an extra space 🤦‍♂️

**Review by MoLow — APPROVED** [review]

**Comment by MoLow** [comment]
@mdkitzman thanks for your contribution and sorry for all the friction!
